### DISCUSSION TITLE
Query related to make and break the circuit?

### DISCUSSION THREAD

**Discussion by rana253** [CURENT/andes#340] [Modeling]
Hi Dr. Cui,

Is there any model related to make and break the circuit in real time? eg. Circuit breaker, switch 
Your response in this regards will be very helpful.

Thank You!!

**Comment by cuihantao** [comment]
Is the model Toggler what you are looking for?

Regards,
Hantao Cui

**Reply by rana253** [reply]
Hi Dr. Cui,
Thank you for your response.
Model Toggler is operated at a predefine time. 
I want to connect `Circuit Breaker or switch` between resource and associated bus like (between battery-bus ,PV-bus, load-bus and Genrou-bus). So can I isolate the resources dynamically when it is not required in the system and restore it when it is required in the system.
Is their any model like Circuit Breaker and Switch? If not, is their any other way to do it in real time.

**Comment by cuihantao** [comment]
To disconnect a device at any time, you can alter its connectivity status “u” to zero. You can  implement your logic using a perturbation file.

Regards,
Hantao Cui

**Reply by rana253** [reply]
Hi Dr. Cui,

Thank you for your response.
As per your suggestion I will implement the logic by using a perturbation file.
Just a small clarification: can Line be  disconnected by altering connectivity status “u” to zero?

Regards,
Rana Pratap

**Comment by cuihantao** [comment]
Yes. See https://docs.andes.app/en/latest/getting_started/formats/xlsx.html#common-parameters

Regards,
Hantao Cui

**Comment by rana253** [comment]
Hi Dr. Cui,

Thank you for your response.

I have implemented following two logics by using perturbation file to disconnect the Line_0.
1. ` system.Line.alter('u', 'Line_0', 0)`
2.  `system.Line.u.v[0] = 0`


But it gives following errors on running the simulation:

**Jacobian matrix is singular.
Suspect diagonal elements: [27, 38]
  0%|                                                    | 0/100 [00:00<?, ?%/s]**

**Time step reduced to zero. Convergence is not likely.
Simulation terminated at t=0.0000 s.
Simulation completed in 0.0050 seconds.
TimeSeries does not contain any time stamp.**

Please guide me to do in correct way.

Regards,
Rana Pratap

**Reply by cuihantao** [reply]
It's possible that the line you disconnected triggered the isolation of a bus or buses. A connectivity check is needed after altering the line status. It can be done by telling the system that a custom event has been set. 

```python

def pert(t, system):

    if (your logic triggers a line trip):
        system.Line.alter('u', 'Line_0', 0)
        system.TDS.custom_event = True
```

**Reply by rana253** [reply]
Hi Dr. Cui,

Thank you for your response.
I have tried with this "system.TDS.custom_event = True" but our system is not allowing to do this because of isolation of bus.

Now I am trying to weaken the `Line_0` by increasing its resistance and reactance to the maximum possible extent ( altering `bhk` and `ghk`). Can I go with this method to achieve our requirement?
Can you guide me in what ratio the alteration of `bhk` and `ghk` should be done in order to make the line weak.

Please suggest me if there is any other possible way to do this? 

Regards,
Rana Pratap

**Reply by cuihantao** [reply]
> I have tried with this "system.TDS.custom_event = True" but our system is not allowing to do this because of isolation of bus.

Not sure why. `system.TDS.custom_event = True` is meant to fix singular Jacobians caused by system separation.

**Comment by rana253** [comment]
Hi Dr. Cui,

Thank you for your response.  
I have considered `1-genrou,1-pvd1,2-esd1 and 6-loads`. Genrou is connected with `SLACK` bus-1 and I want to `isolate` Genrou and run the system in `islanded` mode. 
Can you guide me, what changes to be done in order to simulate the system in `islanded` mode?

**Reply by cuihantao** [reply]
Can you upload an established test case and the perturbation file that you've tried?

**Comment by rana253** [comment]
Hi Dr. Cui,

I wanted to know that how we can switch a slack bus during simulation.
Eg- In system, `Genrou`(slack generator) is connected to Bus 1(slack bus) and `ESD1` is connected to Bus 11. Now I want to convert slack from Bus 1 to Bus 11 (`ESD1`) during simulation. 
Kindly advise how this can be done?

Regards,
Rana Pratap

**Reply by cuihantao** [reply]
There is some confusion about how the static generators are used. The conversion between a static generator, say a Slack or a PV, and a dynamic generator, like `ESD1`, is done at the initialization of time-domain simulations. Once completed, the initialized dynamic devices will replace the static ones, but the unreplaced static generators will remain. There isn't a way to convert it on the fly during the simulation.

One approach to implementing the described procedure is to replace the Slack by a ESD1 at the initialization phase. By nature, the ESD1 needs to be initialized at the same power outputs for power flow consistency. What is the goal of switching a Slack to an ESD1 during a simulation?

**Reply by rana253** [reply]
Our objective is to simulate the disconnection of slack bus `generator` from a system that only has a generator, an ESD1, a PVD1 and load. After the disconnection of the generator, we want to assign `ESD1` as the slack bus.

**Reply by cuihantao** [reply]
Please understand that "slack" generator is only a concept introduced to set the reference voltage and angle and balance any power mismatches. A slack generator strictly regulates the voltage and angle and thus has independent power variables. An ESD1 cannot be a slack generator as its power outputs are subject to controller actions and thus are not independent. 

I'm still not sure of the application background of your scenario. But disconnecting a slack bus generator can be roughly equivalent to disconnecting a negative load with the same amount of power.
